core/impl-base/src/main/java/org/jboss/arquillian/core/impl/loadable/JavaSPIExtensionLoader.java

```diff
@@ -61,23 +72,119 @@
+
+               for (Map.Entry<Object, Object> entry : entries)
+               {
+                  String service = (String) entry.getKey();
+                  String serviceImpls = (String) entry.getValue();
+
+                  addVetoedClasses(service, serviceImpls, classLoader, vetoed);
+
+               }
+
+            }
+            finally
+            {
+               if (inStream != null)
+               {
+                  inStream.close();
+               }
+            }
+         }
+      } catch (IOException e) {
+         throw new RuntimeException("Could not load exclusions from " + EXCLUSIONS, e);
+      }
+
+      return vetoed;
+   }
+
    //-------------------------------------------------------------------------------------||
    // Internal Helper Methods - Service Loading ------------------------------------------||
    //-------------------------------------------------------------------------------------||
 
-   private <T> Set<Class<? extends T>> load(Class<T> serviceClass, ClassLoader loader) 
+   private void addVetoedClasses(String serviceName, String serviceImpls, ClassLoader classLoader, Map<Class<?>, Set<Class<?>>> vetoed)
+   {
+       try
+       {
+          final Class<?> serviceClass = classLoader.loadClass(serviceName);
+          final Set<Class<?>> classes = loadVetoedServiceImpl(serviceImpls, classLoader);
+
+          final Set<Class<?>> registeredVetoedClasses = vetoed.get(serviceClass);
+          if (registeredVetoedClasses == null) {
+             vetoed.put(serviceClass, classes);
+          }
+          else
+          {
+             registeredVetoedClasses.addAll(classes);
+          }
+
+       }
+       catch (ClassNotFoundException e)
+       {
+           // ignores since this is a veto that it is not applicable
+       }
+   }
+
+   private Set<Class<?>> loadVetoedServiceImpl(String serviceImpls, ClassLoader classLoader)
+   {
+
+      final StringTokenizer serviceImplsSeparator = new StringTokenizer(serviceImpls, ",");
+      final Set<Class<?>> serviceImplsClass = new LinkedHashSet<Class<?>>();
+
+      while (serviceImplsSeparator.hasMoreTokens())
+      {
+         try
+         {
+            serviceImplsClass.add(classLoader.loadClass(serviceImplsSeparator.nextToken().trim()));
+         }
+         catch (ClassNotFoundException e)
+         {
+            // ignores since this is a veto that it is not applicable
+         }
+      }
+
+      return serviceImplsClass;
+   }
+
+   private <T> Set<Class<? extends T>> load(Class<T> serviceClass, ClassLoader loader)
    {
       String serviceFile = SERVICES + "/" + serviceClass.getName();
 
-      LinkedHashSet<Class<? extends T>> providers = new LinkedHashSet<Class<? extends T>>();      
-      LinkedHashSet<Class<? extends T>> vetoedProviders = new LinkedHashSet<Class<? extends T>>();      
-      
+      LinkedHashSet<Class<? extends T>> providers = new LinkedHashSet<Class<? extends T>>();
```

Comment:
I understand, but what is the benefit? Except of illusionary performance gain (only valid on Dalvik I guess).